syntax = "proto3";

package signals;

option go_package = ".sil";

service Signals {
  rpc EnumerateSignals (EnumerateSignalsRequest) returns (EnumerateSignalsResponse) {}
  rpc WriteSignal(WriteSignalRequest) returns (WriteSignalResponse) {}
  rpc ReadSignal(ReadSignalRequest) returns (ReadSignalResponse) {}
}

enum SignalType {
    SIGNAL_TYPE_UNKNOWN = 0;
    SIGNAL_TYPE_ADC = 1;
    SIGNAL_TYPE_DIGITAL = 2;
    SIGNAL_TYPE_PWM = 3;
}

message AdcSignal {
    double voltage = 1;
}

message PwmSignal {
    uint32 duty_cycle = 1;
    uint32 frequency = 2;
}

message DigitalSignal {
    bool level = 1;
}

message SignalInfo {
    string ecu_name = 1;
    string signal_name = 2;
    SignalType signal_type = 3;
}

message EnumerateSignalsRequest {}

message EnumerateSignalsResponse {
  bool status = 1;
  string error = 2;
  repeated SignalInfo signals = 3;
}

message WriteSignalRequest {
    string ecu_name = 1;
    string signal_name = 2;
    oneof value {
        AdcSignal value_adc = 3;
        DigitalSignal value_digital = 4;
        PwmSignal value_pwm = 5;
    }
}

message WriteSignalResponse {
    bool status = 1;
    string error = 2;
}

message ReadSignalRequest {
    string ecu_name = 1;
    string signal_name = 2;
}

message ReadSignalResponse {
    bool status = 1;
    string error = 2;
    oneof value {
        AdcSignal value_adc = 3;
        DigitalSignal value_digital = 4;
        PwmSignal value_pwm = 5;
    }
}
